docs/user/build-customization.rst

```diff
@@ -104,6 +105,61 @@ To avoid this, it's possible to unshallow the clone done by Read the Docs:
          - git fetch --unshallow
 
 
+Skip build based on a condition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There may be situations where you want to skip a build that was automatically triggered when someone on your team pushed to the repository.
+Skipping builds will allow you to speed up review times and also help us reduce server costs and ultimately our environmental footprint.
+Consider the following scenarios:
+
+* the build depends on an external situation that's not met yet
+* there were no changes on the documentation files
+
+In these scenarios, you can skip the build by executing a custom command that checks for that particular condition
+and exits with code ``439`` to skip it, or ``0`` to continue building the documentation normally.
+If any of the commands return this particular exit code,
+Read the Docs will stop the build immediately,
+mark it as "Cancelled",
+and communicate to your Git platform (GitHub/GitLab) that the build succeeded (green tick ✅) so the pull request is in a mergeable state.
+
+Here is an example that skip build from pull requests when there are no changes to the ``docs/`` folder compared to the ``origin/main`` branch:
+
+.. code-block:: yaml
+   :caption: .readthedocs.yaml
+
+   version: 2
+   build:
+     os: "ubuntu-22.04"
+     tools:
+       python: "3.11"
+     jobs:
+       post_checkout:
+         # Skip building pull requests when there aren't changed in the docs directory.
+         # `--quiet` exits with a 1 when there **are** changes,
+         # so we invert the logic with a !
+         #
+         # If there are no changes (exit 0) we force the command to return with 439.
+         # This is a special exit code on Read the Docs that will cancel the build immediately.
+         - if [ $READTHEDOCS_VERSION_TYPE = "external" ]; then ! git diff --quiet origin/main -- docs/ && exit 439; fi
+
+
+This other example shows how to skip a build if the commit message contains ``skip ci`` on it:
+
+.. code-block:: yaml
+   :caption: .readthedocs.yaml
+
+   version: 2
+   build:
+     os: "ubuntu-22.04"
+     tools:
+       python: "3.11"
+     jobs:
+       post_checkout:
+         # Use `git log` to check if the latest commit contains "skip ci",
+         # in that case exit the command with 439 to skip the build
+         - case `git --no-pager log --pretty="tformat:%s" -1` in *"skip ci"*) exit 439;; *);; esac
```

Comment:
Hmm. I implemented this:

https://github.com/spacetelescope/jdaviz/commit/b15517753139916ed5598ef1871fa955f12da8d4

And then I tried using the directive in https://github.com/spacetelescope/jdaviz/pull/1829 which created https://readthedocs.org/projects/jdaviz/builds/18586511/ that did not get cancelled.

What did I do wrong?

Reply:
That's because that `git log` command only outputs the _title_ of the commit, without the description. Checking over the description as well, may require a different `--pretty=` format argument.

Reply:
In that case, you may want something like:

```
$ git --no-pager log --pretty="tformat:%s -- %b" -1
Use multi-line bash examples -- They are easier to read
```

Reply:
note the `%s` _and_ `%b`

I found this at https://git-scm.com/docs/pretty-formats

Reply:
I think that works. Thanks! https://readthedocs.org/projects/jdaviz/builds/18586938/

However, would be nice if such checks is done before cancelling a previous job from the same branch. Consider this use case:

1. Commit A pushed with real changes. It is still running the RTD build.
2. Commit B is trivial and has skip directive in its commit message. It is pushed while the run for Commit A is still not finished.

Now, RTD will cancel the run for Commit A and then fail the build for Commit B (because of the skip). Commit A should keep running if Commit B is skipped.

Reply:
Gotcha! This would be pretty hard to implement with the current code and structure.

If you want this to be considered for a next iteration of this feature, please open a new issue explaining these details 👍

You have shared great feedback already! 💯

Reply:
Actually, Read the Docs know nothing about the reason of the build being skipped, so I don't think it will be able to make these decisions 🤷‍♂️

Reply:
Well, in other CI services, skip directives (e.g., `[ci skip]` in commit message) are defined internally, so I guess that would make things easier. For example, Actions doesn't even start when it detects the skip (for better or worse), hence avoid cancelling the last job even if auto-cancel is defined in its workflow.

So, should I still bother to open a new issue? Or maybe at least document the canceling behavior in the doc?

Reply:
The cancelling behavior is already documented in another PR (I'm on mobile, but you can quickly find it)

The "skip ci" behaviour from other CIs is exactly what you need for this use case. However, this feature of using exit codes to get the outcome is not exactly that feature, but am approximation.

It's not integrated into the CI itself, but using another feature (the one implemented on this PR) to have a similar behavior.

Because it's not implemented at the core of the CI we can't handle these particular cases.

We should handle it in a different way, by exploring the commit message from the webhook itself and decide to no trigger the build at all.

What I'm describing is a different feature from what this PR implements. A valid feature, tho.

Reply:
BTW, this is the commit that documents "When to cancel builds" and explains how it works: [`111ee35` (#9649)](https://github.com/readthedocs/readthedocs.org/pull/9649/commits/111ee350847d5ce3857a24078a7b21692bd950fd)

Reply:
@humitos let's say that a build is automatically cancelled and superseded by a build that's then also cancelled.

Is there a button on auto-cancelled builds that can force them to run, i.e. this one?

![image](https://user-images.githubusercontent.com/374612/201088670-74c08773-635d-4e1e-9f98-18c5566d0881.png)

(sorry, I couldn't quickly find an auto-cancelled build)

Reply:
Yeah, we could change the logic behind that button if we want to.

https://github.com/readthedocs/readthedocs.org/blob/eeb62b71afde2d865cbb59145ebfd7b0500c799c/readthedocs/templates/builds/build_detail.html#L80-L82

However, if I'm reading that comment correctly, I think it should be present in the case that @pllim was describing.

Reply:
Not autocancelling is better but I guess a button to force rebuild is better than nothing. Thanks!

So, I should have been using exit code 183 and not 439, even after this PR is merged?

Reply:
The exit code is the same. POSIX automatically calculate the module of integers bigger than 255

Reply:
I guess it works. 😸 https://readthedocs.org/projects/jdaviz/builds/18597605/